README.md
# DTest
Test and call function with random predicted arguments.
`DTest` generates random arguments and call the function with the arguments.
Generating argument is easy and simple, you can customize it anytime.

```js
var users = [];
function NewUser(uid, password) {
  users.push({ uid, password });
}

var test = DTest(NewUser, [{
  type: DTest.TYPE_NUMBER,
  config: [0, 99999]
}, {
  type: DTest.TYPE_STRING
}], (func, args, res) => {
  console.log("Calling");
  res(func.apply({}, args));
  console.log("Done");
});
console.log(test);
```


<!--
# Development

If you want to test your function quickly, view [DTest Development](https://dlvdls18.github.io/DTest/dev.html).
-->



# Getting Started

## Installation

Download the file [dtest.js](https://cdn.jsdelivr.net/gh/dlvdls18/DTest@main/src/dtest.js)
then add the following code:

```html
<script src="path/to/dtest.js"></script>
```

Or use JSDelivr CDN:
```html
<script src="https://cdn.jsdelivr.net/gh/dlvdls18/DTest@main/src/dtest.js"></script>
```

## Building Your Project

1. Enter these commands in your terminal

```bash
mkdir MyProject
cd MyProject
touch index.html
touch index.js
```

2. Open your favorite IDE and open the file `/MyProject/index.html`
3. Copy this html template and paste it to `index.html`



```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <title>MyProject</title>
  </head>
  <body>
    <p>Hello, World!</p>
  </body>
  <!-- <script src="path/to/dtest.js"></script> -->
  <script src="https://cdn.jsdelivr.net/gh/dlvdls18/DTest@main/dtest.js"></script>
  <script>
    var users = [];
    function NewUser(uid, password) {
      users.push({ uid, password });
    }
    var test = DTest(NewUser, [{
      type: DTest.TYPE_NUMBER,
      config: [0, 99999]
    }, {
      type: DTest.TYPE_STRING
    }], (func, args, res) => {
      console.log("Calling");
      res(func.apply({}, args));
      console.log("Done");
    });
    console.log(test);
  </script>
</html>
```

4. Start Coding :tada:



# Documentations

## Basics

These are the arguments of `DTest` function:

|  Name   |     Type     |       Description      | Required |
|---------|--------------|------------------------|----------|
| func    | Function     | Function to test       |   Yes    |
| rules   | Array Object | How arguments generate |   Yes    |
| handler | Function     | How function called    |   No     |
| config  | Object       | Each arguments config  |   No     |


## Writing Rules

### Features

`Rules` are important for generating random arguments.
`Rules` can:
- Customize the argument type
  - String
  - Number
    - Can customize the range (0-10 by default)
  - Boolean
  - Any (String, Number, Boolean)
  - Array String
    - Customized array size (required)
  - Array Number
    - Customized array size (required)
    - Can customize the range (0-10 by default)
  - Array Boolean
    - Customized array size (required)
  - Any Array (String, Number, Boolean)
    - Customized array size (required)
  - Object (Value of String)
    - Customized object size (required)
    - Keys can be String or Number
  - Object (Value of Number)
    - Customized object size (required)
    - Keys can be String or Number
    - Can customize the range (0-10 by default)
  - Object (Value of Boolean)
    - Customized object size (required)
    - Keys can be String or Number
  - Object (Any Value)
    - Customized object size (required)
    - Keys can be String or Number
- Add predicted value
  - Set the argument value to given predicted value instead of randomized value

More features coming soon!

### Format

```js
[{
  type: number,
  config: array
}, ...]
```

#### Types

- String `DTest.TYPE_STRING`
- Number `DTest.TYPE_NUMBER`
- Boolean `DTest.TYPE_BOOLEAN`
- Any `DTest.TYPE_ANY`
- Array String `DTest.TYPE_ARRAY_STRING`
- Array Number `DTest.TYPE_ARRAY_NUMBER`
- Array Boolean `DTest.TYPE_ARRAY_BOOLEAN`
- Array Any `DTest.TYPE_ARRAY_ANY`
- Object String `DTest.TYPE_OBJECT_STRING`
- Object Number `DTest.TYPE_OBJECT_NUMBER`
- Object Boolean`DTest.TYPE_OBJECT_BOOLEAN`
- Object Any `DTest.TYPE_OBJECT_ANY`

#### Configuration

Each types are number from 0 to 11.
For selecting the type, I use 
```js
var rule = rules[i];
var type_functions = [string, randint, boolean, ...]
var type_number = rule.type;
type_functions[type_number].apply({}, rule.config || []);
```

for quick development instead of
```js
var rule = rules[i];
if(rule == 0) ...
else if(rule == 1) ...
else if(rule == 2) ...
...
```

or

```js
var rule = rules[i];
switch(rule.type) {
  case 0: ...
  break;
  case 1: ...
  break;
  case 3: ...
  break;
  case ...: ...
  break;
  default: ...
  break;
}
```


So `config` is an array for arguments.
These are the arguments for `type functions`:


|           Type              |      Arguments       |                                             Description                                                   |
|-----------------------------|----------------------|-----------------------------------------------------------------------------------------------------------|
|          String             |       Format         | String Format                                                                                             |
|          Number             |      Min, Max        | Randomization Range (default 0-10)                                                                        |
|          Boolean            | T Percent, F Percent | True and False Percentage                                                                                 |
| Array String, Boolean, Any  |        Size          | Length of the array (required)                                                                            |
|       Array Number          |   Size, Min, Max     | Length of the array (required), Randomization Range (default 0-10)                                        |
| Object String, Boolean, Any |      Size, Key       | Length of the object (required), Key Type (number) (String or Number)                                     |
|      Object Number          | Size, Min, Max, Key  | Length of the object (required), Randomization Range (default 0-10), Key Type (number) (String or Number) |


## Argument Length

Since the format of `rules` is array object.
You can use many argument input as you want.

### Example

#### Code


```js
function add() {
  var num = arguments[0];
  for(var i = 1; i < (arguments.length + 1); i++) {
    num += arguments[i];
  }
  return num;
}
```

```js
DTest(add, [{
  type: DTest.TYPE_NUMBER,
  config: [1, 100]
}, {
  type: DTest.TYPE_NUMBER,
  config: [1, 100]
}, {
  type: DTest.TYPE_NUMBER,
  config: [1, 100]
}, ...]);
```


#### Expected Output
```
Example Random Number Value:
arg1 = 10
arg2 = 3
arg3 = 8
arg4 = 5
```

```
Output:
> 26
```


## Customizing How Function Called

You can customize how function is called with the argument `handler`.
`handler` argument accept a function with arguments

- func
- args
- res



### Func

Function to Test

### Args

Generated Arguments

### Res

A function to end the promise and return the result.

```js
res( /*  Return Value Of "func"  */ );

// Recommend Use
res( func.apply({}, args) );
```

Since `args` is array type. It is recommended to use `func.apply` function to call the function.


## Return Value

### With and Without Handler

### Without Handler

```js
{
  call_args: array,
  func: function,
  result: any,
  recall: function,
  recall_promise: function
}
```


### With Handler

With `handler`, return value is a `Promise`.

```js
DTest(...).then(function(result) {
  console.log("Test Ended");
  console.log(result);
});
```

You don't need to add `.catch` function since there will be no errors unless:

- Some required arguments/values are missing
- Incorrect argument/value type
- Function to test throws an error


The `promise` cannot be ended until the function `res` is called.

After the `promise` ended, it will call `resolve` (which is `.then`) with 1 argument:

```js
{
  call_args: array,
  func: function,
  result: any,
  recall: function,
  recall_promise: function
}
```


### Recall and Recall Promise


#### Recall

If you want to call the function again with promise and the same arguments,
You can use `.recall` function


#### Recall Promise

If you want to call the function again with the same arguments,
You can use `.recall_promise` function.

This function requires 1 argument, `handler`.
`handler` feature is the same for `test with handler`.


#### Return Value

```js
{
  call_args: array,
  func: function,
  result: any
}
```

## String Format

Feature added at **1.1**


If you want simple string, you can use:

- `DTest.FORMAT_STRING`
- `DTest.FORMAT_NUMBER`

For argument of `string`

```js
string(...);
```


### String

#### Before

```
aOnT05jS6A
```

#### After

```
Helamety Pode
```


### Number

#### Before

```
j9bwm7F
```

#### After

```
9217 7271 1072 8157
```


### How It Made

#### String

Function `generateword` generates a consonant and a vowel

```
Example:
Ha
Le
Ko
We
Bu
```

Function `generatefullword` call `generateword` in random times and capitalize each words

```
Example:
Hakoremu Pomelari
Lore No
Duqolema Pire Gomiso
```


#### Number

Just like `String`, It generates random number (1000-9999) for random times

```
Example:
8352 3815 7243 7268 1638
8363 5273 9485 
9372 8989 3632 8263
```

## DTest Configuration

Feature added at **1.1**

```js
{
  // remove argument duplicates
  remove_duplicates: boolean (false by default)
}
```

More configuration will added soon

## Any... Configuration

Feature added at **1.1**

Function `any` `anyarray` `anyobject` doesn't have any arguments to customize each types.
To keep predictable, I added an argument called `config` which is very useful for type arguments.

Since `config` in `rules` are for arguments, `config` for these functions are object.
These are available keys for the `config`:
- string
- randint (number)
- bool (boolean)


```js
{
  string: [DTest.FORMAT_STRING],
  randint: [1, 1000],
  bool: [DTest.PERCENTAGE_LOW, DTest.PERCENTAGE_NORMAL]
}
```


## Boolean Percentage

Feature added at **1.1**

You can control how random boolean generates with this feature


### Arguments

- True Percentage
- False Percentage




### Percentages

- `DTest.PERCENTAGE_LOW`
- `DTest.PERCENTAGE_NORMAL`
- `DTest.PERCENTAGE_HIGH`


### Example

```js
DTest(..., [{
  type: DTest.TYPE_BOOLEAN,
  config: [DTest.PERCENTAGE_LOW, DTest.PERCENTAGE_HIGH]
}]);
```


info/HISTORY.md
# History


## First Release

June 2, 2022

[View at GitHub](https://github.com/dlvdls18/DTest/)

[README.md](https://github.com/dlvdls18/DTest/blob/main/README.md)

[Main JS File](https://github.com/dlvdls18/DTest/blob/main/src/dtest.js)

### Source released at GitHub

View at GitHub



### `README.md` updated

View at GitHub



### File `CONTRIBUTION.md` added

View at GitHub



### File `TODO.md` added

View at GitHub



### File `version.txt` added

View at GitHub



### Documentation removed at `src/dtest.js`

For neat comtent



### TODO removed at `README.md`

For neat content



***

## First Update

June 2, 2022

[View at GitHub](https://github.com/dlvdls18/DTest/)

[README.md](https://github.com/dlvdls18/DTest/blob/main/README.md)

[Main JS File](https://github.com/dlvdls18/DTest/blob/main/src/dtest.js)

### File `version.txt` updated

View at GitHub


### File `TODO.md` updated

View at GitHub



### File `info/HISTORY.md` added

View at GitHub


### Added range to `array any` and `array object` for number

Task from `README.md`



### Static variable `FOMRAT_STRING`, `FORMAT_NUMBER`, `PERCENTAGE_LOW`, `PERCENTAGE_NORMAL`, `PERCENTAGE_HIGH`, `VERSION` added

Task from `README.md`



### Make the static variable unwritable

Just like constant.
Task from `README.md`



### Remove duplicates added

Task from `README.md`



### Function `DTest` argument added `config`

For customizable argument



### Function `any` argument added `config`

For customizable argument



### Function `anyobject` argument added `config`

For customizable argument



### Function `anyarray` argument added `config`

For customizable argument



### Code `tf[tn].apply({}, ...)` changed to `tf[tn].apply({}, config[tf[tn].name] || []`

For each argument configuration



### String format added

Task from `README.md`



### Nullable added

Task from `README.md`



### Predicate value added

Task from `README.md`



### Boolean percentage added

Task from `README.md`



### Task `Add arguments verification` added to `TODO.md`

Incase the passed arguments are wrong which caused errors.



### Task `Transfer version.txt, TODO.md, CONTRIBUTION.md to info/...` added to `TODO.md`

For neat file list



### Files `TODO.md` and `CONTRIBUTION.md` transferred to `info/...`

For neat file list



### File `version.txt` deleted

Since `DTest.VERSION` is defined, I deleted the file `version.txt` because I'm struggling to manually edit the file.



### Description added for each history commit

For reasons and explanations.


src/dtest.js
/**
 * DTest
 * License: MIT
 * (c) 2022 - dlvdls18
 * https://github.com/dlvdls18/DTest/
 */

function DTest(func, rules, handler, config) {
  // important note:
  // every type function values must be exactly
  // string, randint, bool, any, stringarray, numarray, boolarray, anyarray, stringobject, numobject, boolobject, anyobject
  var inc_num = -1;
  // increment the number and return it
  function inc() {
    inc_num++;
    return inc_num;
  }
  // return random number with range
  function randint(min, max) {
    min = min || 0
    max = max || 10;
    return Math.floor(Math.random() * (max + min + 1)) + min;
  }
  // return random string
  function string(format) {
    var chars = Array.from("ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789");
    var random = "";
    for(var i = 0; i < randint(3, 20); i++) random += chars[randint(0, chars.length - 1)];
    var fmt = null;
    if(format != null) {
      fmt = "";
      if(format == 0) {
        for(var i = 0; i < randint(1, 3); i++) fmt += " " + generatefullword();
      } else {
        for(var i = 0; i < randint(2, 10); i++) fmt += " " + randint(1000, 9999);
      }
      fmt = fmt.slice(1);
    }
    return fmt || random;
  }
  // return random boolean
  function bool(tp, fp) {
    if(tp == null && fp == null) return randint(0, 1) == 1;
    else return percentage(tp, fp);
  }
  // return random any type
  function any(config) {
    var tf = [string, randint, bool];
    var tn = randint(0, 2);
    return tf[tn].apply({}, config ? config[tf[tn].name] || [] : []);
  }
  // return random number array
  function numarray(size, min, max) {
    var arr = [];
    for(var i = 0; i < size; i++) arr.push(randint(min, max));
    return arr;
  }
  // return random string array
  function stringarray(size) {
    var arr = [];
    for(var i = 0; i < size; i++) arr.push(string());
    return arr;
  }
  // return random boolean array
  function boolarray(size) {
    var arr = [];
    for(var i = 0; i < size; i++) arr.push(bool());
    return arr;
  }
  // return an array with random type
  function anyarray(size, config) {
    var arr = [];
    for(var i = 0; i < size; i++) {
      var t = randint(0, 2);
      var f = [string, randint, bool];
      arr.push(f[t].apply({}, config ? config[f[t].name] || [] : []));
    }
    return arr;
  }
  // return a string object
  function stringobject(size, key) {
    key = key || 0;
    inc_num = -1;
    var kf = [string, inc];
    var obj = {};
    for(var i = 0; i < size; i++) {
      var kt = arraysearch(kf, key);
      obj[kt()] = string();
    }
    return obj;
  }
  // return a number object
  function numobject(size, min, max, key) {
    key = key || 0;
    inc_num = -1;
    var kf = [string, inc];
    var obj = {};
    for(var i = 0; i < size; i++) {
      var kt = arraysearch(kf, key);
      obj[kt()] = randint(min, max);
    }
    return obj;
  }
  // return a boolean object
  function boolobject(size, key) {
    key = key || 0;
    inc_num = -1;
    var kf = [string, inc];
    var obj = {};
    for(var i = 0; i < size; i++) {
      var kt = arraysearch(kf, key);
      obj[kt()] = bool();
    }
    return obj;
  }
  // return an object with random type
  function anyobject(size, key, config) {
    key = key || 0;
    inc_num = -1;
    var kf = [string, inc];
    var obj = {};
    for(var i = 0; i < size; i++) {
      var kt = arraysearch(kf, key);
      var t = randint(0, 2);
      var f = [string, randint, bool];
      obj[kt()] = f[t].apply({}, config ? config[f[t].name] || [] : []);
    }
    return obj;
  }
  // return the array item using absolute out of range position value
  function arraysearch(array, position) {
    var pos = 0;
    for(var i = 0; i < Math.abs(position); i++) {
      if(pos >= (array.length - 1)) pos = -1;
      pos++;
    }
    return array[pos];
  }
  // capitalize each words
  function title(str) {
    var res = "";
    for(var i = 0; i < str.split(" ").length; i++) {
      var w = str.split(" ")[i];
      res += " " + w.charAt(0).toUpperCase() + w.slice(1);
    }
    return res.slice(1);
  }
  // generate consonant and vowel string
  function generateword() {
    // some letters are removed for stable word
    var vowels = Array.from("aeiou");
    var consonants = Array.from("bdfgklmnprst");
    return consonants[randint(0, consonants.length - 1)] + vowels[randint(0, vowels.length - 1)];
  }
  // generate full un/defined word
  function generatefullword() {
    var gen = "";
    for(var i = 0; i < randint(1, 10); i++) gen += generateword();
    return title(gen);
  }
  // percentage boolean
  function percentage(y, n) {
    if(y == n) return randint(0, 1) == 0;
    else {
      var num = randint(0, 5);
      if(num < (y + 1)) return true;
      else if(num < (n + 1)) return false;
      else return y > n;
    }
  }
  // convert set to array
  function set_to_array(set) {
    var arr = [];
    set.forEach(function(item) { arr.push(item) });
    return arr;
  }
  // loop each rules
  var call_args = [];
  for(var i = 0; i < rules.length; i++) {
    var rule = rules[i];
    var rf = [string, randint, bool, any, stringarray, numarray, boolarray, anyarray, stringobject, numobject, boolobject, anyobject];
    var rt = rule.type;
    var rfu = null;
    var b = bool();
    if(typeof rt == "number") rfu = arraysearch(rf, rt);
    else rfu = arraysearch(rf, rt[randint(0, rt.length - 1)]);
    if(rule.nullable == true && b == true) call_args.push(null);
    else call_args.push(rule.pred_value || rfu.apply({}, rule.config || []));
  }
  // remove duplicates
  if(config && config.remove_duplicates == true) {
    call_args = new Set(call_args);
    call_args = set_to_array(call_args);
  }
  // call the function directly
  if(handler == null) return {
    call_args, func,
    result: func.apply({}, call_args),
    recall: function() {
      return {
        call_args: this.call_args,
        func: this.func,
        result: this.func.apply({}, this.call_args),
        self: this
      }
    },
    recall_promise: function(handler) {
      return new Promise(function(resolve, reject) {
        var that = this;
        handler.call({}, this.func, this.call_args, function(result) {
          resolve({ result, call_args: that.call_args, handler});
        });
      });
    }
  }
  // call the function with customized argument passing
  else return new Promise(function(resolve, reject) {
    handler.call({}, func, call_args, function(result) {
      resolve({ result, call_args, func, handler});
    });
  });
}

DTest.TYPE_STRING = 0;
DTest.TYPE_NUMBER = 1;
DTest.TYPE_BOOLEAN = 2;
DTest.TYPE_ANY = 3;
DTest.TYPE_ARRAY_STRING = 4;
DTest.TYPE_ARRAY_NUMBER = 5;
DTest.TYPE_ARRAY_BOOLEAN = 6;
DTest.TYPE_ARRAY_ANY = 7;
DTest.TYPE_OBJECT_STRING = 8;
DTest.TYPE_OBJECT_NUMBER = 9;
DTest.TYPE_OBJECT_BOOLEAN = 10;
DTest.TYPE_OBJECT_ANY = 11;

DTest.FORMAT_STRING = 0;
DTest.FORMAT_NUMBER = 1;

DTest.PERCENTAGE_LOW = 0;
DTest.PERCENTAGE_NORMAL = 1;
DTest.PERCENTAGE_HIGH = 2;

DTest.VERSION = 1.1;

// make all static variable unwritable
for(var i in DTest) {
  if(["TYPE", "FORMAT", "PERCENTAGE", "VERSION"].includes(i.split("_")[0])) {
    Object.defineProperty(DTest, i, {
      writable: false
    });
  }
}